Feature/10ns increments

vsmar opened the pull request:
Added 10ns resolution for the PulseBlaster.
Ensures that a programmed pulse sequence is valid, otherwise issuing an error.

Main error statements improved:
- Invalid pulse timing parameter: indicates the conflicting duration or start time, and that it is not a 10ns multiple or >50ns.
- Invalid pulse sequence: Occurs when multiple pulse edges fall within<50ns. Error indicates conflicted channels and the time at which the channel conflicts.

vsmar committed:
10ns increment + validation update

vsmar committed:
removed unnecessary import

mfparsons314 commented on pulseblaster/PBInd.py:
```diff
@@ -4,54 +4,57 @@
+            - Pulses can not be less than 50ns apart (across all channels)
+            - Puses can not overlap by less than 50ns
     '''
     def __init__(self,
```
For every user facing method, we need to have a restructured text docstring.  (e.g. see below)  You can setup your IDE (at least in pycharm) to generate a template for you automatically as an autocomplete.  I didn't initially write this code, so this practice was not followed, but I'd like you to implement this.

"""
This is a reST style.

:param param1: this is a first param
:param param2: this is a second param
:returns: this is a description of what is returned
:raises keyError: raises an exception
"""

mfparsons314 commented on pulseblaster/PBInd.py:
```diff
@@ -4,54 +4,57 @@
-
         self.spinapi = spincore_spinapi
 
     def on(self, pin, start, length):
```
add reST docstring

mfparsons314 commented on pulseblaster/PBInd.py:
```diff
@@ -4,54 +4,57 @@
         self.set(pin, start, length, 0)
 
-
     def make_clock(self, pin, period):
```
add reST docstring

mfparsons314 commented on pulseblaster/PBInd.py:
```diff
@@ -60,15 +63,12 @@ def make_clock(self, pin, period):
 
-
-
     def program(self, offsets, loops):
```
add reST docstring

mfparsons314 commented on pulseblaster/PBInd.py:
```diff
@@ -125,20 +128,12 @@ def print_instructions(self):
-                index = d
-                return index
-
     def write_instruction(self, actual_chs, loops):
```
add reST docstring

mfparsons314 commented on pulseblaster/PBInd.py:
```diff
@@ -200,7 +195,7 @@ def write_instruction(self, actual_chs, loops):
     def get_offset_chs(self, offsets):
```
add reST docstring

mfparsons314 commented on pulseblaster/PBInd.py:
```diff
@@ -223,6 +218,11 @@ def get_offset_chs(self, offsets):
         return offset_chs, front_ind, back_ind
 
     def set(self, pin, start, len, val):
```
add reST docstring

mfparsons314 commented on pulseblaster/PBInd.py:
```diff
@@ -263,3 +263,70 @@ def get_state(self, state_index, state_array, num_elements):
 
             return stateslice
+
+    def validate_pulse_instructions(self, chs_matrix):
```
add reST docstring

mfparsons314 commented on pulseblaster/README.md:
Please add a couple of paragraphs description in the readme about how your validation works.  Describe very briefly the problem with the old code and then in more detail how you fixed it.  If there are any under-the-hood changes that get made to timing sequences without explicitly notifying the user, then please describe them clearly here.

vsmar replied:
Thanks for the review! 
I implemented the following changes:
- Added reST style comments to all the public-facing functions, changed private functions to have a preceding underscore. 
- Removed the depricated offset functionality (added a note about it in the README)
- Updated the README with an overview of the functionality and changes, and other key information. 
- Added an example to illustrate a new use-case it can handle, and updated the other examples and the toml file.

Let me know if there are any remaining concerns!

mfparsons314 reviewed (changes requested):
Hi Victor,

Overall this looks pretty good to me, except that the documentation needs to be improved.  I realize you were following the practices of this initial code, but I didn't write it and the original documentation was not up to my standards for user facility code.  Please add better documentation to each method.

I scanned through the logic of you validation method.  I think it looks ok, but let's spend 5 minutes talking about it if you are in today's project meeting.

vsmar committed:
Updated docstring comments to follow reST structure, and provided a detailed README.md

vsmar committed:
Updated readme, examples and toml file

vsmar commented:
Hi Max (@mfparsons314 ),

I didn't notice that the previous comments we're now outdated due to the file changes, until I submitted my comment.
I made the changes you outline, let me know if anything else is needed!

> Thanks for the review! I implemented the following changes:
> 
> * Added reST style comments to all the public-facing functions, changed private functions to have a preceding underscore.
> * Removed the depricated offset functionality (added a note about it in the README)
> * Updated the README with an overview of the functionality and changes, and other key information.
> * Added an example to illustrate a new use-case it can handle, and updated the other examples and the toml file.
> 
> Let me know if there are any remaining concerns!

mfparsons314 reviewed (approved):
Looks good - thanks for the updates!
